### PR TITLE
[WIP] What: Cloud Native Guide Posts Issue #961

### PR DESCRIPTION
**Pull request by TheFoxAtWork**
Why:

* Capture milestones from successful projects
* provide guidance and recommendations to projects seeking graduation
* Provide criteria to guide TOC on determining when sandbox projects should depart from sandbox

This change addresses the need by:

* Modifying archiving.md to reflect Sandbox Departure link
* Modify sandbox.md to set Sandbox Departure criteria
* Modify README.md to capture Milestones link
* create project milestones for community review, improvement, and feedback

Signed-off-by: Emily Fox <themoxiefoxatwork@gmail.com>

### CODE REVIEW

File: process/archiving.md [hunk 1]
```diff
@@ -10,7 +10,7 @@ When voting on a proposal to archive a project, TOC members may wish to consider
 
 To archive a project:
 
-* A proposal must be put forth to the TOC repo
+* A proposal must be put forth to the TOC repo in the form of an issue
```
Comment (xmulligan) [inline process/archiving.md:13]:
```suggestion
* A proposal must be put forth to the TOC repo in the [form of an issue](https://github.com/cncf/toc/issues/new?assignees=caniszczyk%2C+amye%2C+idvoretskyi%2C+jeefy&labels=archive&template=archived.md&title=%5BARCHIVE%5D+project)
```

File: process/sandbox.md [hunk 2]
```diff
@@ -91,3 +91,16 @@ Frequency of reviews can be found on the [repo's README under "Frequency"](https
 ### Annual review
 
 Once in the Sandbox, projects are subject to an [Annual Review](https://github.com/cncf/toc/blob/master/process/sandbox-annual-review.md).
+
+### Sandbox departures
```
Comment (xmulligan) [inline process/sandbox.md:95]: Should this be called "sandbox archival" just to keep wording consistent and not introduce another term that needs to be understood by everyone?

File: process/project_milestones.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+* Project holds community meetings or active discussions
+  * Zoom with meeting notes occurring more than twice a quarter (or community.cncf.io!) 
+  * Use of GitHub discussions, issues, or a group messaging service (slack, wechat, etc.)
+* Project guides community contributions
+  * PRs from community members receive timely feedback and discussion from the maintainers
+* Project works with an angel adopter
+  * Partners closely with the project to address organizational needs of the project early on
+* Project begins solidifying a versioning schema and release cadence
+  * A roadmap clearly defines initiatives that align with releases
+  * Releases begin to have establish a pattern of regularity
```
Comment (xmulligan) [inline process/project_milestones.md:19]:
```suggestion
  * Releases begin to establish a pattern of regularity
```

File: process/project_milestones.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+* Project has engaged a Technical Advisory Group for feedback on some portion of the project relevant to the domain
```
Comment (xmulligan) [inline process/project_milestones.md:20]:
```suggestion
* Project has engaged a [Technical Advisory Group](/tags/README.md) for feedback on some portion of the project relevant to the domain
```

File: process/project_milestones.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+* Project has a clearly discoverable governance doc that covers the basics of decision making to include the commit bit
```
Comment (xmulligan) [inline process/project_milestones.md:21]:
```suggestion
* Project has a clearly discoverable governance doc that covers the basics of decision making and how to earn the commit bit
```

File: process/project_milestones.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+* TBD - add more!
+
+## Incubation
+
+Incubation milestones are guiding points to highlight success areas a maturing project can accomplish before Graduation. Not all Incubating projects will achieve every milestone, however many successful graduated projects have demonstrated these milestones prior to their Graduation. How each project achieves some or all of these milestones will vary widely. Projects do not need to meet this in order to graduate, rather these assist projects in meeting the Graduation criteria.
+* Project has a robust set of governance documentation that defines how the direction of the project is decided upon and managed which considers company diversity in decision making
```
Comment (xmulligan) [inline process/project_milestones.md:27]:
```suggestion
* Project has a robust set of governance documentation that defines how the direction of the project is decided upon and managed which considers company diversity in decision making. TAG Contributor Strategy has many great [governance templates](https://github.com/cncf/project-template)
```

File: process/project_milestones.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+* Project can demonstrate its application, practice, and adjustments to its governance documentation as a result of regular project operations
+* Project has a robust and mature security posture as evidenced by
+  * Threat model
+  * [Joint review](https://github.com/cncf/tag-security/tree/main/assessments#components-of-the-security-review-package)
+  * First passing then Silver OpenSSF Best Practice Badge
```
Comment (xmulligan) [inline process/project_milestones.md:32]: This is currently different than graduation requirements. I would keep it consistent across both places https://github.com/cncf/toc/blob/main/process/graduation_criteria.md#graduation-stage

File: process/project_milestones.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+* Project has a robust set of governance documentation that defines how the direction of the project is decided upon and managed which considers company diversity in decision making
+* Project can demonstrate its application, practice, and adjustments to its governance documentation as a result of regular project operations
+* Project has a robust and mature security posture as evidenced by
+  * Threat model
+  * [Joint review](https://github.com/cncf/tag-security/tree/main/assessments#components-of-the-security-review-package)
+  * First passing then Silver OpenSSF Best Practice Badge 
+* Project documentation includes getting started guides, operating/administration instruction, security call-outs, and other core elements necessary to ease adoption.
+* Project provides scalability benchmarks and sets expectations around scaling
```
Comment (xmulligan) [inline process/project_milestones.md:34]: This is unclear to me. Do we have examples or resources we could provide here?

File: process/project_milestones.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+* Project captures use cases to showcase how it can be best used to solve common problems.
+  * These may leverage observations and concerns from adopters
+* Project contribution and community documentation clearly defines expectations when contributing (to include non-code contributions).
+* Project defines a process to “build the bench” of maintainers and project leaders such as a Contribution Ladder which considers succession planning
```
Comment (xmulligan) [inline process/project_milestones.md:38]:
```suggestion
* Project defines a process to “build the bench” of maintainers and project leaders such as a Contribution Ladder which considers succession planning and getting maintainers from multiple organizations
```

File: process/project_milestones.md [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+* Project guides community contributions
+  * PRs from community members receive timely feedback and discussion from the maintainers
+* Project works with an angel adopter
+  * Partners closely with the project to address organizational needs of the project early on
+* Project begins solidifying a versioning schema and release cadence
+  * A roadmap clearly defines initiatives that align with releases
+  * Releases begin to have establish a pattern of regularity
+* Project has engaged a Technical Advisory Group for feedback on some portion of the project relevant to the domain
+* Project has a clearly discoverable governance doc that covers the basics of decision making to include the commit bit
+* TBD - add more!
```
Comment (xmulligan) [inline process/project_milestones.md:22]:
```suggestion
* Project has an Adopters file and starts capturing public production users and use cases
* TBD - add more!
```

**Comment by lizrice** [comment]
Preparing for graduation, we (Cilium) have been getting recommendations and requirements to do a lot of security-related audits and assessments:

* 3rd party security audit - [clear graduation requirement](https://github.com/cncf/toc/blob/main/process/graduation_criteria.md)
* [Open SSF (aka Core Infrastructure)](https://bestpractices.coreinfrastructure.org/en) best practices badge - clear graduation requirement
* TAG Security review [self assessment](https://github.com/cncf/tag-security/blob/main/assessments/guide/self-assessment.md) and/or [joint review](https://github.com/cncf/tag-security/blob/main/assessments/guide/joint-review.md) - not documented requirements. Seems a duplication of effort for the project security folks who are already assisting the 3rd party audit. 
* [CLOmonitor badge](https://github.com/cncf/clomonitor) - includes some security-related checks, recommended by CNCF
* [OpenSSF scorecard](https://securityscorecards.dev/) - recommended by 3rd party audit, overlaps with CLOmonitor
* [SLSA assessment](https://slsa.dev/) - recommended by 3rd party audit

They’re all good things on their own, but taken together they’re repetitive and overlapping, and are causing quite a lot of frustration ("we're drowning in checklists and paperwork".)  It would be really helpful to have a clear steer on which of these are actually required.

**Comment by TheFoxAtWork** [comment]
There's a lot here we can get better clarity on.

The TAG Security Review/Assessment process was heavily reworked a few years ago to structure it such that projects may complete portions of the overall "Security Package" aligned with where they are at in the CNCF "lifecycle". This revision was designed to prevent the security review (which often doesn't have enough reviewers and as a result does take longer) from becoming a bottleneck for a project pursuing graduation. If projects completed the "self-assessment" during Sandbox, then optional "joint-review" during Incubation (well before considering graduation) these two security documents form a security package that can be leveraged by Security Auditors to assist in the formal audit. The joint-review builds on top of the self-assessment, but both documents do NOT replace a Security Audit, while joint-reviews do have some areas of overlap with Security Audits, the Security Audits go more in depth in other areas. They should be viewed as cumulatively complementary to one another, however a joint-review is NOT a requirement for graduation (nor is the self-assessment for incubation).

I am not comfortable placing TAG Security and the joint-reviews in a position where all projects come to them b/c they are now the executors of a graduation requirement.  It's not feasible with the amount of projects in the landscape and the amount of security reviewers we have.

There as a been a lot of evolution in open source and supply chain security these past few years, with increasing maturity of various assessments that look at different portions of a project - from its architecture to its repository configuration and release processes. We've not yet revisited how these changes should be weighed or considered in the CNCF lifecycle and I believe this is worth pursuing by community through the Security TAG.

File: process/project_milestones.md (hunk 7 above)
Comment (TheFoxAtWork) [reply process/project_milestones.md:32]: @xmulligan it is different from graduation requirements. Perhaps this is better suited as a split? Achieve Passing during incubation (to align with graduation) and then Silver post graduation?  these are not requirements and are optional for project (except where they exist as requirements). Silver has (IMO) reasonable and (in all likelihood)  a fair amount of requirements beyond passing that are already met by highly mature projects.

What do you think of this change? breaking it out and moving Silver to after graduation?

File: process/project_milestones.md (hunk 8 above)
Comment (TheFoxAtWork) [reply process/project_milestones.md:34]: This would be a new discipline (and should include performance with scalability) - there are a few projects that have done comparisons of performance between themselves and other projects and products.  This would drop the comparison portion and rather focus on the recommended architecture, a few deployment options (single cluster, multi-cluster, multi-cloud), how it scales (increasing nodes, increased traffic, load, or processes, expected completion time, etc.) It may also be as simple as expressing the CPU utilization of running the project depending on various configurations.  The goal would be to set expectations with potential users/adopters that may also be beneficial in understand if the project is performing optimally in their environment or if there are other factors contributing to performance or scaling degradation.

File: process/project_milestones.md (hunk 7 above)
Comment (tomkerkhove) [reply process/project_milestones.md:32]: As a maintainer, I think this makes sense to still have "goals"/"points for improvement" after graduating as that is not the end of the governance story for a project

File: process/project_milestones.md (hunk 8 above)
Comment (tomkerkhove) [reply process/project_milestones.md:34]: As a maintainer, I believe it's good to also define when this is _not_ a requirement. For example, KEDA handles autoscaling but there is no aspect of scalability/performance benchmark as we're just building on top of KEDA with an additional operator.

You can always make the case that this is still required; but for some projects I believe it is less beneficial or the impact on a project would be too big.

Again for KEDA, we'd need to spin up big (expensive) clusters with sample workloads (1000+) to do benchmarking of something that is just built on top of Kubernetes' building blocks (HPA / metric servers)

**Comment by tomkerkhove** [comment]
> Preparing for graduation, we (Cilium) have been getting recommendations and requirements to do a lot of security-related audits and assessments:
> 
> * 3rd party security audit - [clear graduation requirement](https://github.com/cncf/toc/blob/main/process/graduation_criteria.md?rgh-link-date=2023-01-27T12%3A44%3A13Z)
> * [Open SSF (aka Core Infrastructure)](https://bestpractices.coreinfrastructure.org/en) best practices badge - clear graduation requirement
> * TAG Security review [self assessment](https://github.com/cncf/tag-security/blob/main/assessments/guide/self-assessment.md?rgh-link-date=2023-01-27T12%3A44%3A13Z) and/or [joint review](https://github.com/cncf/tag-security/blob/main/assessments/guide/joint-review.md?rgh-link-date=2023-01-27T12%3A44%3A13Z) - not documented requirements. Seems a duplication of effort for the project security folks who are already assisting the 3rd party audit.
> * [CLOmonitor badge](https://github.com/cncf/clomonitor) - includes some security-related checks, recommended by CNCF
> * [OpenSSF scorecard](https://securityscorecards.dev/) - recommended by 3rd party audit, overlaps with CLOmonitor
> * [SLSA assessment](https://slsa.dev/) - recommended by 3rd party audit
> 
> They’re all good things on their own, but taken together they’re repetitive and overlapping, and are causing quite a lot of frustration ("we're drowning in checklists and paperwork".) It would be really helpful to have a clear steer on which of these are actually required.

Agreed

**Comment by tomkerkhove** [comment]
Some practical things that I think would be nice to include are:
- The fact that new proposals should end up in CNCF-based Google workspace (see https://github.com/cncf/toc/issues/980)
- If we should use issues vs pull requests to open proposals. If we stick with pull requests - What should be in them? Because typically people include a whole proposal in the PR which then gets ignored/must be ported over to the DD doc which costs a lot of time
  - Been through this twice to write a full proposal (which takes a lot of time) only to see it get ignored and need to create a Google doc afterwards

**Comment by TheFoxAtWork** [comment]
I had an excellent discussion with a community member yesterday about the amount of security asks of projects and the current throughput of Security TAG in reviews, specifically regarding the joint-reviews.  
It was recommended that we extract the threat modeling process (when its ready - this is still in development) from the joint review and focus on this for projects immediately after moving to incubation.  The threat modeling portion is designed to take up 2-3 hours of a projects time and is likely useful for projects immediately following incubation.  

What do others think about this as a _milestone_?

**Comment by dims** [comment]
+1 @TheFoxAtWork

File: process/README.md [hunk 11]
```diff
@@ -127,5 +127,6 @@ Additionally, the TOC might recommend that you apply for Incubation stage. This
 
 It is fine for a project to stay in the Sandbox indefinitely while it is still active, but if a project has genuinely stalled we can save everyone’s effort by archiving it.
 
+# Maturity resources for projects at all levels
 
-
+The [CNCF Technical Advisory Groups](/tags/README.md) have a wide variety of resources available to assist projects in building communities, securing their projects, and applying best practices across a variety of domains like networking, app-delivery, and much more! In addition to these resources, the TOC has as collection of [Cloud Native Milestones](project_milestones.md) which projects may leverage as guiding points along their journey — from sandbox to beyond graduation.
```
Comment (craigbox) [inline process/README.md:132]:
```suggestion
The [CNCF Technical Advisory Groups](/tags/README.md) have a wide variety of resources available to assist projects in building communities, securing their projects, and applying best practices across a variety of domains like networking, app-delivery, and much more! In addition to these resources, the TOC has a collection of [Cloud Native Milestones](project_milestones.md) which projects may leverage as guiding points along their journey — from sandbox to beyond graduation.
```

File: process/project_milestones.md (hunk 5 above)
Comment (craigbox) [reply process/project_milestones.md:21]: "the commit bit" is probably not a well understood term.  I don't know if the CNCF has a defined term (["maintainer" is definitely overloaded](https://github.com/cncf/foundation/issues/433)), but I think "how to earn permission to approve pull requests" or something similar would help.

(I assume the SCM for CNCF projects does not _have_ to include GitHub and a more inclusive terms than Pull Requests might be appropriate)

**Comment by craigbox** [comment]
I would caution that even though it might not be the stated intention, these may well end up interpreted as "unwritten criteria" for moving between maturity levels, especially in light of #992.

File: process/project_milestones.md (hunk 5 above)
Comment (lizrice) [reply process/project_milestones.md:21]: I've been thinking about Roadmaps. In the past I've been in favour of them as a requirement but I think I'm changing my mind because I'm not convinced they are as valuable as we'd like to think. 

Background / food for thought: the Linux kernel does not have a published roadmap. Everyone involved knows what's going through mailing list discussions.

When we say things like a roadmap that's aligned with releases, combined with a regular cadence of releases, this feels dangerously close to requiring open source projects to make commitments that end users might come to think they can rely on.

The main concern I have is that it doesn't seem very aligned with the ethos of open source to require developers to sign up to fixed commitments. Sure, in our community most people are being paid to work on projects but not everyone is. And if someone from Company A makes a commitment to deliver a particular feature in a certain release, but then Company A decides they should spend their time elsewhere, who from the project is supposed to pick up the slack to make sure the commitment is reached? 

It might work for very large projects, but the more I think about it, the more it seems like a recipe for burnout amongst contributors, and disappointment amongst end users. 

Projects generally use issues to communicate what features and fixes people are working on. A feature that meets some set of criteria (up to the project) makes it into the next release. Isn't that sufficient? 

And if we require more, isn't that rather overstepping the idea that we don't impose any particular governance on projects?

Comment (monadic) [reply process/project_milestones.md:21]: +100 - we really need to stop turning CNCF into a giant bureaucracy that treats all projects the same.  This kills innovation stone dead and will put so many people off.  Also, please stop imposing requirements on project maintainers that add a lot of work and will lead to burnout, without a very strong compensating reason and balance

Comment (TheFoxAtWork) [reply process/project_milestones.md:21]: @lizrice Thanks for bringing this up, I dont disagree. We currently have roadmaps integrated through various portions of the lifecycle so we'll need to evaluate what the value is of these in both understanding the project's longer goals as well as the direction they are heading, however they are always point in time for evaluations along the lifecycle. 

Your point on perceived commitment is spot on as well, and I'm not confident "messaging" here will resolve this disconnect. We'll need to consider more about how we can view the direction of the project at an evaluation, without dictating how the project expresses that.

Comment (TheFoxAtWork) [reply process/project_milestones.md:21]: @monadic Thank you for the feedback.  This PR on milestones does not establish new or additional requirements on projects, it is designed to serve as ` guiding points along their [project's] journey — from sandbox to beyond graduation` which they `may leverage`.  However, as @craigbox mentioned later, these may be seen as requirements. This is something will need to rework in this PR to avoid.
Regarding burnout, agreed. The CNCF lifecycle is structured to evaluate projects for stability, maturity, and adoption potential (to varying degrees) over the years, as we've learned more about project success, we're beginning to distill steps projects have taken along their path to success. Not as requirements, more as a catalog of what has and hasnt worked and more context around why.  

Would providing more reasoning/clarity/what-it-does assist with the milestones? Such that as a project maintainer looking to move to incubation from sandbox (for instance) has ample information about how these may be beneficial   to decide if documenting a roadmap is worthwhile for their project, or perhaps including a vision/direction, or neither, works for their needs?

File: process/project_milestones.md (hunk 10 above)
Comment (TheFoxAtWork) [reply process/project_milestones.md:22]: Including this but adjusting language to reflect more project driven implementation with examples of "how"

**Comment by TheFoxAtWork** [comment]
> The fact that new proposals should end up in CNCF-based Google workspace (see https://github.com/cncf/toc/issues/980)
@tomkerkhove this would be best in the graduation criteria/process documentation as this was a recent change.  @amye is this something you can ensure is integrated?

> Been through this twice to write a full proposal (which takes a lot of time) only to see it get ignored and need to create a Google doc afterwards

Completely agree this is time consuming and inconsistent. @amye whatever updates we make for this should reduce duplicative effort.  We the template for graduation with a lot of information in it, but perhaps its worth figuring out where this goes in a DD or a DD refresh versus initiating a process.

File: process/project_milestones.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
+# Cloud Native Milestones
+
+Milestones are intended to outline checkpoints for projects, that have assisted past projects as they grew and mature in the cloud native ecosystem. They are NOT requirements for moving between levels, please refer to [the graduation criteria](https://github.com/cncf/toc/blob/main/process/graduation_criteria.md) for requirements.
```
Comment (leecalcote) [inline process/project_milestones.md:3]:
```suggestion
Milestones are intended to outline checkpoints for projects, that have assisted past projects as they grew and matured in the cloud native ecosystem. They are NOT requirements for moving between levels, please refer to [the graduation criteria](https://github.com/cncf/toc/blob/main/process/graduation_criteria.md) for requirements. 
```

Suggesting past tense for "mature" considering the sentence's past tense of "grew".

**Comment by TheFoxAtWork** [comment]
Per discussion on today's TOC/TAG Call here is a google doc of the primary changes for better discussion/facilitation of this PR: https://docs.google.com/document/d/1aL0mxInyz-qtRT2QM7kdr9riM36uihILtgM4TR35cPQ/edit?usp=sharing

File: process/project_milestones.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+  * PRs from community members receive timely feedback and discussion from the maintainers
+* Project works with an angel adopter
+  * Partners closely with the project to address organizational needs of the project early on
+* Project has a second organization contributing
+  * May begin path to maintainership
+* Project begins solidifying a versioning schema and release cadence
+  * A roadmap or other document clearly defines project direct or initiatives that may align with releases
+  * Releases begin to establish a pattern of regularity
+* Project has engaged a [Technical Advisory Group](/tags/README.md) for feedback on some portion of the project relevant to the domain
+* Project has a clearly discoverable governance doc that covers the basics of decision making and how to earn permissions to approve pull request, sometimes referred to as "the commit bit"
```
Comment (jberkus) [inline process/project_milestones.md:25]:
```suggestion
* Project has a clearly discoverable governance doc that covers the basics of decision making, including who approves PRs and how committers/maintainers are chosen
```

File: process/project_milestones.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+* Project has an angel adopter that assists in stability and production runtime or deployment
+* Project has 3 angel adopters
+  * May be from different industry verticals
```
Comment (jberkus) [inline process/project_milestones.md:28]: Not quite understanding the language around "angel adopters", can this explain better?

File: process/project_milestones.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+  * May have differing use cases
+* TBD - add more!
+
+## Incubation
+
+Incubation milestones are guiding points to highlight success areas a maturing project can accomplish before Graduation. They are not requirements. Not all Incubating projects will achieve every milestone, however many successful graduated projects have demonstrated these milestones prior to their Graduation. How each project achieves some or all of these milestones will vary widely. Projects do not need to meet this in order to graduate, rather these assist projects in meeting the Graduation criteria.
+* Project has a robust set of governance documentation that defines how the direction of the project is decided upon and managed which considers company diversity in decision making. [TAG Contributor Strategy](https://github.com/cncf/tag-contributor-strategy) has many great [governance templates](https://github.com/cncf/project-template).
```
Comment (jberkus) [inline process/project_milestones.md:35]:
```suggestion
* Project is evolving and developing a robust set of governance documentation that defines how the project is managed, and its roadmap is determined, which considers maintainer turnover and company diversity. [TAG Contributor Strategy](https://github.com/cncf/tag-contributor-strategy) has many great [governance templates](https://github.com/cncf/project-template).
```

File: process/project_milestones.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+* Project refreshes their roadmap planning with a reasonable cadence
+* Project has engaged with a few Technical Advisory Groups to ensure they are considering best practices across multiple technical and governance facets
+* Project has defined a process for managing the conduct of its community, either establishing its own group to do this, leveraging the existing leadership, or deferring to the CNCF
+* Project maintains a dependency graph so adopters fully understand the complexity and risk of use
+* TBD - add more!
+
+## Graduation
+
+Graduation milestones are guiding points to highlight continuing success areas for highly mature project after they graduate. They are not requirements. Not all Graduated projects will achieve or perform these milestones, however many robust and mature projects will adopt these milestones as "regular business" to maintain the health and growth of the project and its community.
+* Project reviews it's governance process annually for improvements to improve clarity, consistency, and inclusivity
```
Comment (jberkus) [inline process/project_milestones.md:58]:
```suggestion
* Project regularly updates its governance documentation to accurately reflect changes to processes, staffing, roles, and activities
```

We're already giving Incubating projects the full milestone of "complete governance", which means that all we can ask graduated projects to do is to keep that up to date.  No?

File: process/project_milestones.md [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+* Project has a well established and exercised contributor growth ladder or other community construct to "build the bench" of maintainers or project leaders
```
Comment (jberkus) [inline process/project_milestones.md:59]: We already ask for this at incubating.  Would it be worth asking a project to have demonstrated turnover?

File: process/sandbox.md [hunk 18]
```diff
@@ -91,3 +91,16 @@ Frequency of reviews can be found on the [repo's README under "Frequency"](https
 ### Annual review
 
 Once in the Sandbox, projects are subject to an [Annual Review](https://github.com/cncf/toc/blob/master/process/sandbox-annual-review.md).
+
+### Sandbox archival
```
Comment (jberkus) [inline process/sandbox.md:95]: I feel like this should be a separate PR; folks are going to have feelings about archival, and I'd like to merge the milestones while that argument plays out.

**Review by jberkus — CHANGES_REQUESTED** [review]

File: process/project_milestones.md (hunk 17 above)
Comment (TheFoxAtWork) [reply process/project_milestones.md:66]: Done: https://github.com/cncf/toc/pull/1020

File: process/project_milestones.md (hunk 5 above)
Comment (jberkus) [reply process/project_milestones.md]: I've suggested different language for this point.

File: process/project_milestones.md [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+Some milestones may not apply to all projects and should be leveraged as guiding points. Milestones are described as the desired state with examples for how to achieve this. They are not requirements.
+* Project has the basics of security underway
+  * Completion of the [self-assessment](https://github.com/cncf/tag-security/blob/main/assessments/guide/self-assessment.md)
+* Project holds community meetings or active discussions
+  * Zoom with meeting notes occurring more than twice a quarter (or community.cncf.io!) 
+  * Use of GitHub discussions, issues, or a group messaging service (slack, wechat, etc.)
+* Project guides community contributions
+  * PRs from community members receive timely feedback and discussion from the maintainers
+* Project works with an angel adopter
```
Comment (craigbox) [inline process/project_milestones.md:17]: Duplicated at line 26?

File: process/project_milestones.md [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+* Project is evolving and developing a robust set of governance documentation that defines how the project is managed, and its roadmap is determined, which considers maintainer turnover and company diversity. [TAG Contributor Strategy](https://github.com/cncf/tag-contributor-strategy) has many great [governance templates](https://github.com/cncf/project-template).
+* Project can demonstrate its application, practice, and adjustments to its governance documentation as a result of regular project operations
+* Project has a robust and mature security posture which could be achieved by one or more of the following:
+  * Threat model
+  * [Joint review](https://github.com/cncf/tag-security/tree/main/assessments#components-of-the-security-review-package)
+  * [Passing level](https://bestpractices.coreinfrastructure.org/en/criteria/0) of the [OpenSSF Best Practice Badge](https://bestpractices.coreinfrastructure.org/)
+    * Note: Completion and maintenance of **a** OpenSSF Best Practice Badge is [a requirement for graduation](https://github.com/cncf/toc/blob/main/process/graduation_criteria.md#graduation-stage)
```
Comment (craigbox) [inline process/project_milestones.md:45]: `**an**` 😉

File: process/project_milestones.md [hunk 21]
```diff
@@ -57,6 +55,9 @@ Incubation milestones are guiding points to highlight success areas a maturing p
 * Project has engaged with a few Technical Advisory Groups to ensure they are considering best practices across multiple technical and governance facets
 * Project has defined a process for managing the conduct of its community, either establishing its own group to do this, leveraging the existing leadership, or deferring to the CNCF
 * Project maintains a dependency graph so adopters fully understand the complexity and risk of use
+* Project has 3 or more adopters using the project in production to support completion of Due Diligence and adopter interviews
+* Project plans activities to increase project adoption trajectory 3 adopters
```
Comment (lizrice) [inline process/project_milestones.md:59]:
```suggestion
* Project plans activities to increase project adoption 
```

File: process/project_milestones.md [hunk 22]
```diff
@@ -57,6 +55,9 @@ Incubation milestones are guiding points to highlight success areas a maturing p
...
+* Project has more than 5 regular contributors, more than half of which are from different organizations
```
Comment (lizrice) [inline process/project_milestones.md:60]: This one worries me A LOT. It clearly incentivizes projects and companies to do the wrong thing, in order to meet these "milestones".  

For example, fearing that they are being seen to have "too many" contributions going into the open source project, the major backer of a project decides to hold their engineering efforts back into closed-source add-ons. Or use outsourcing companies and partners rather than hire people to work full-time on the project. Or decide not to contribute a project to the CNCF because it's going to be such an uphill struggle for the project to see itself as successful according to these criteria. 

None of these outcomes are good for anyone in our community. 

(Also, why arbitrarily 5? And what does "regular" mean? We could spend time debating these points - but that is time that could be better spent on actually progressing / supporting projects.)

Comment (TheFoxAtWork) [reply process/project_milestones.md:60]: These are good counterpoints @lizrice . We also don't want to have a single entity (company or organization or individual) be the majority decision maker of a projects technical direction, if that entity is acquired or disappears, the project can be left out to dry, and in some case unable to recover because the primary pool of contribution has evaporated.

@raravena80 @lizrice - How do we strike a balance in these milestones — as guiding points projects _could_ leverage to in order to become more resilient, mature, and widely adopted versions — that could help projects avoid the scenarios you present as well as a increase the likelihood of project's survivability with primary entity's complete absence?  What _actionable_ goals/ideals/structure/governance could be provided for projects to increase their resilience and survival likelihood in the event a core contributing entity ceases to be or continue? 

Ricardo mentioned open and flexible language here, would the following adjustments meet the intent?

```suggestion
* Project has multiple contributors who regularly participate in the growth and development of the project in accordance with the projects synchronous and asynchronous cadence
  * Project contribution governance defines timeliness and methods for recovering roles and permissions of contributors that drop from regular participation as _defined by the project_ (roles such as issue triage, community management, PR reviewer, maintainer, release lead, etc. and regularity such as participation in two of three releases, facilitation of 4 community meetups annually, etc.)
```

**Comment by monadic** [comment]
Hi, I say this with care and respect for the intentions of the people on this thread.  If the CNCF wants to continue to be a home for innovative and leading projects, then it needs to listen to people who work on them and make them successful.  In my view, the kind of rules being proposed here will greatly narrow what sorts of projects can thrive in CNCF and make it ultimately unappealing.  Please can we arrange a TOC call with appropriate representation to discuss.

-alexis

**Comment by TheFoxAtWork** [comment]
@monadic thank you for engaging, we can certainly schedule this for a discussion on a TOC call!  It is apparent that the guiding points identified here are still being viewed as rules, requirements, or other forms of constraint which is not what they should be used for.  

I would love to hear more about your experience to see which, if any, of these describe the path for moving levels of the projects you've been involved in. Scaling that experience and making it approachable for other projects — if they would like to leverage it — would be beneficial for new projects and maybe even existing projects seeking to grow and move levels.  When we have a date, I'll update here.  Thank you again!

File: process/project_milestones.md (hunk 22 above)
Comment (jberkus) [reply process/project_milestones.md:60]: I can't say I'm keen on removing specifics from this.  Folks in the community manager role for a project like having clear targets.  Once a project gets to applying Graduated, if they don't have at least five regular contributors, the application isn't likely to go well.  

Note that these are milestones, not updates to the specific requirements.

**Comment by monadic** [comment]
thanks @TheFoxAtWork 

this is something we spent a lot of time thinking about when creating CNCF - what makes a successful OSS project?  the answer is that they come from different places but a consistent theme is small companies innovating ** and sustaining ** them.  this has numerous consequences.  I love multipolar projects too.

**Comment by TheFoxAtWork** [comment]
Looks like this actually already on the Agenda for the meeting next week (march 21st) @monadic (apologies for missing it earlier)

File: process/project_milestones.md (hunk 22 above)
Comment (raravena80) [reply process/project_milestones.md:60]: 👍  I believe all these can be covered in the [March 21st TOC meeting](https://docs.google.com/document/d/1jpoKT12jf2jTf-2EJSAl4iTdA7Aoj_uiI19qIaECNFc/edit#heading=h.8q8nrqluehva).

**Comment by monadic** [comment]
8am PT?  please can you advertise on toc & alums email lists

**Comment by craigbox** [comment]
> It is apparent that the guiding points identified here are still being viewed as rules, requirements, or other forms of constraint which is not what they should be used for.

The points under discussion today _are_ being discussed as a criteria for graduation in #992.  [I've commented at length there](https://github.com/cncf/toc/issues/992#issuecomment-1474704202).

(I would love to be involved in an in-person discussion, but not enough to get up at 4am I'm afraid.)

**Comment by lizrice** [comment]
I have some concrete ideas for this milestones document, to de-couple the advice on growing a project from the TOC’s assessment of its maturity.  

* Make this document about _guidance to projects on how to grow_. Some indication of what success looks like could be helpful, but be careful to make it applicable to all projects (so, no arbitrary numbers.)  
* Remove all mention of the sandbox / incubation / graduation levels so that people aren’t tempted to draw their own conclusions about whether they should be considered by the TOC when moving levels.
* Remove the points in this document that appear as guidance to the TOC on how to assess projects. Anything the TOC requires from all projects at a particular maturity level should be documented in the criteria, and anything else should be a matter of judgement. 

I’m aiming to be there for the discussion coming up in this week’s TOC meeting - looking forward to it. 
—

> Folks in the community manager role for a project like having clear targets. 

As someone with experience on both the TOC and in community management for Cilium, I can tell you that no-one wants yet more checklists where it’s unclear how important they are - our experience with the [myriad security assessments](https://github.com/cncf/toc/pull/997#issuecomment-1406461915) advised to us during the course of our graduation application is an example. If something looks like it _might_ be considered, projects feel obliged to document them in their DD just in case it affects their application. And (sadly) people with some axe to grind will bring up the lack of some specific “milestone” during the public comment period. Neither of these outcomes are remotely helpful to the project or the user community - they just create additional work for the projects and for the TOC.  

> Once a project gets to applying Graduated, if they don't have at least five regular contributors, the application isn't likely to go well.

Most Sandbox projects have at least five regular contributors, so this number isn’t helpful as a target for graduation. What’s the right number? For massive (Kubernetes-esque) projects it’s probably one or two orders higher than this; for a small but critical security library, five might be totally appropriate. 

There is no “one size fits all” set of measurements that is appropriate for the needs of all projects. That’s why levelling assessments can’t be boiled down to a set of checkboxes, and it’s why previous incarnations of the TOC have deliberately resisted setting arbitrary thresholds, in favour of exercising judgement on a per-project basis.

**Comment by JustinCappos** [comment]
I'll also mention a few more boundary cases so they are on everyone's radar.  Specifically, that it's hard to assess projects which are more specification focused given guidelines which are more often to assess an implementation.

For example, when considering a specification project, should you count maintainers / approvers for the specification as those controlling entities?  Should this also include the implementation maintainers?

A healthy rate of churn for a successful and active specification may only involve a few changes a year.

What if there are different compliant implementations from different groups that individually do not have a high degree of maintainer diversity?  Would this be considered reasonable diversity, given one can always pick which group's implementation to use?

**Review by nikhita — APPROVED** [review]
PR has been open for a while. LGTM, let's :shipit: 

Having guide posts will help with evaluating sandbox annual reviews and providing a structured way to help sandbox projects figure out a growth path.

**Comment by TheFoxAtWork** [comment]
PR needs refreshed to align with recent changes in the TOC repo. Will update later.
